### DISCUSSION TITLE
Getting Method threw 'java.util.NoSuchElementException' exception.

### DISCUSSION THREAD

**Discussion by sharma-kunal** [JanusGraph/janusgraph#4032] [Q&A]
I have my Janusgraph Schema defined like this,

```
        management.makePropertyKey("name").dataType(String.class).make();
        management.makePropertyKey("p_uuid").dataType(String.class).make();
        management.makePropertyKey("site_context").dataType(String.class).make();
        management.makePropertyKey("company_name").dataType(String.class).make();
        management.makePropertyKey("start_date").dataType(Date.class).make();
        management.makePropertyKey("end_date").dataType(Date.class).make();
        management.makePropertyKey("current").dataType(Integer.class).make();
```

```
management.buildIndex("seeker", Vertex.class).addKey(management.getPropertyKey("p_uuid"))
                .indexOnly(management.getVertexLabel("seeker"))
                .buildCompositeIndex();
        management.buildIndex("company", Vertex.class).addKey(management.getPropertyKey("company_name"))
                .indexOnly(management.getVertexLabel("company"))
                .buildCompositeIndex();
```

Now I have inserted data in my graph,

when I query the graph like

```
g.V('<id>').valueMap().next()
```

I get the response
```
p_uuid: <value>
name: <name>
site_context: <site_context>
```

but when I do like this,

```
g.V().hasLabel("seeker").has("p_uuid", <p_uuid>).next()
```

It fails with the error

```
Method threw 'java.util.NoSuchElementException' exception.
```

But if I do

```
g.V().hasLabel("seeker").has("name", <name>).next()
or
g.V().hasLabel("seeker").has("site_context", <site_context>).next()
```

Since it is returning `p_uuid` when using `valueMap()` but when I do I search query with the `has()` method, why does it fails with `'java.util.NoSuchElementException` ? 

What am I missing here ?

**Comment by sheeehy** [comment]
The java.util.NoSuchElementException occurs when the traversal has no elements to provide in the result set. It means that the graph does not contain a vertex with label "seeker" and property "p_uuid" set to the value you provided in <p_uuid>.

Here are some common reasons why you might be running into this issue:

Data Consistency: The p_uuid value you are searching for may not exist in the database at the time of query execution. Double-check the values by executing g.V().hasLabel("seeker").valueMap().toList().

Commit: If you have recently added elements to the graph, make sure that you have committed the transaction to make them visible to queries.

Indexing: Ensure that your index on p_uuid for the label "seeker" is enabled and properly constructed. Indexes sometimes take a while to become effective.

Case Sensitivity: Property keys are usually case-sensitive. Double-check that the case of p_uuid in your has query matches exactly with the one in the graph.

Type Matching: Ensure that the data type of the value you are passing in has("p_uuid", <p_uuid>) is a String, as defined in your schema.

Graph Cache: Sometimes, if there is any kind of caching involved, clearing the cache can help.

**Comment by sharma-kunal** [comment]
I found the issue here. I am not sure if this is an issue with Janusgraph or not, let me specify the details here. 

So I am using Cassandra as my backend. I deleted my keyspace in cassandra using the command

```
drop keyspace
```

After which I ran my code, that created the new keyspace with the same schema that was previously created. 

But when I ran the queries on this new keyspace using Janusgraph, all fields that were part of `composite index` failed with the above error.

The solution was to just restart the Janusgraph server once. 

I am not sure if I should delete the backend storage as I did with the `drop keyspace` command, but if it's the wrong way to delete the backend storage and composite index, then it should be specified in the documentation.
